File: code/239. Sliding Window Maximum/solution.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        if nums == [] or k == 1:
            return nums
        
        deque = [0]
        result = []
        
        for i in range(1, len(nums)):
            while deque != [] and nums[deque[-1]] <= nums[i]:
                del deque[-1]
            deque.append(i)
            if deque[0] < i - k + 1:
                del deque[0]
            if i >= k - 1:
                result.append(deque[0])
        
        result = [nums[i] for i in result]
            
        return result

File: code/239. Sliding Window Maximum/Solution.java
package com.LeetCodeProblems;

public class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        if (nums.length == 0 || k == 1) {
            return nums;
        }

        // initialize
        int len = nums.length;
        int[] result = new int[len - k + 1];
        int[] deque = new int[len];
        int head = 0, tail = -1;
        deque[++tail] = 0;

        // build deque
        for (int i = 1; i < len; i++) {
            while (head <= tail && nums[deque[tail]] < nums[i]) {
                tail--;
            }
            deque[++tail] = i;
            if (deque[head] < i - k + 1) {
                head++;
            }
            if (i >= k - 1) {
                result[i - k + 1] = deque[head];
            }
        }

        // convert index to value
        for (int i = 0; i < len - k + 1; i++) {
            result[i] = nums[result[i]];
        }

        return result;
    }
}
